storage/src/main/java/tech/pegasys/teku/storage/client/CombinedChainDataClient.java

```diff
@@ -535,22 +536,53 @@ public boolean isOptimisticBlock(final Bytes32 blockRoot) {
         .orElse(false);
   }
 
+  private boolean isCanonicalBlockCalculated(
+      final UInt64 slot, final Bytes32 blockRoot, final Bytes32 chainHeadRoot) {
+    return spec.getAncestor(
+            recentChainData.getForkChoiceStrategy().orElseThrow(), chainHeadRoot, slot)
+        .map(ancestor -> ancestor.equals(blockRoot))
+        .orElse(false);
+  }
+
   @SuppressWarnings("unchecked")
-  private SafeFuture<Set<SignedBeaconBlock>> getBlocksByRoots(final Set<Bytes32> blockRoots) {
+  private SafeFuture<Set<BlockAndMetaData>> getBlocksByRoots(
+      final Set<Bytes32> blockRoots, final Bytes32 chainHeadRoot) {
     final SafeFuture<Optional<SignedBeaconBlock>>[] futures =
         blockRoots.stream().map(this::getBlockByBlockRoot).toArray(SafeFuture[]::new);
     return SafeFuture.collectAll(futures)
         .thenApply(
             optionalBlocks ->
-                optionalBlocks.stream().flatMap(Optional::stream).collect(Collectors.toSet()));
+                optionalBlocks.stream()
+                    .flatMap(Optional::stream)
+                    .map(
+                        block ->
+                            toBlockAndMetaData(
+                                block,
+                                isCanonicalBlockCalculated(
+                                    block.getSlot(), block.getRoot(), chainHeadRoot)))
+                    .collect(Collectors.toSet()));
   }
 
-  Set<SignedBeaconBlock> mergeNonCanonicalAndCanonicalBlocks(
+  Set<BlockAndMetaData> mergeNonCanonicalAndCanonicalBlocks(
       final Set<SignedBeaconBlock> signedBeaconBlocks,
       final Optional<SignedBeaconBlock> canonicalBlock) {
     verifyNotNull(signedBeaconBlocks, "Expected empty set but got null");
-    canonicalBlock.ifPresent(signedBeaconBlocks::add);
-    return signedBeaconBlocks;
+    final Set<BlockAndMetaData> result =
+        signedBeaconBlocks.stream()
+            .map(block -> toBlockAndMetaData(block, false))
+            .collect(Collectors.toSet());
+    canonicalBlock.ifPresent(block -> result.add(toBlockAndMetaData(block, true)));
+    return result;
+  }
+
+  private BlockAndMetaData toBlockAndMetaData(
+      final SignedBeaconBlock signedBeaconBlock, final boolean canonical) {
+    return new BlockAndMetaData(
+        signedBeaconBlock,
+        spec.atSlot(signedBeaconBlock.getSlot()).getMilestone(),
+        isOptimisticBlock(signedBeaconBlock.getRoot()),
```

Comment:
This isn't quite right.  We should be marking blocks as optimistic if either the block is optimistic or the chain head is optimistic.